RavenDB-8212 - Killing transformers WIP

ayende opened the pull request:
This is just so we'll have a record on the scope of this change.

ayende committed:
RavenDB-6571
- Adding configuration option for how long to wait before deleting SupersededIndexes (mostly for testing only)
- Unskipping tests

ayende committed:
RavenDB-8212 Kill transformers on server & client (not tests yet)

ayende committed:
RavenDB-8212 - Removing test that check that transformer / index names are unique, it isn't meaningful now

ml054 committed:
RavenDB-8212 Kill transformers

ayende committed:
RavenDB-8212 - Allow to generate raw queries and parameters and use them through the client API

ayende committed:
RavenDB-8212 - Removing tests that verify that transformers work, without actually testing anything interesting there

ayende committed:
RavenDB-8212
- The alternate select location is at the very end of the query, after ORDER BY. This is done so this will match the user expected flow of operations

ayende committed:
RavenDB-8225 Wrong alias can cause strange results

ayende committed:
RavenDB-8212 - Fixing compilation of FastTests

ayende committed:
RavenDB-8228 - Change RQL comment token to // to match the JS comments in declared functions

ayende committed:
RavenDB-8212 - ToListAsync should return List<T>

ayende committed:
RavenDB-8212
- Adding back the extension data for handling of internal @id values (not working yet)
- Will properly send all the document data to the client using declared functions
- Can replace the full document from the query results
- Can pass full document (via alias) to a declared function

ayende committed:
RavenDB-8231 All authentication tests fail if run after midnight, set initial NotBefore to a week ago to avoid it.

ayende committed:
RavenDB-8212 - More removing of transformers from the codebase

ayende committed:
RavenDB-8212 - And more removing of transformers remnants

ayende commented:
Fast test passing, slow tests not compiling yet.

iftahbe committed:
RavenDB-5120 Fix PR comments

ayende committed:
RavenDB-8212 - Better syntax for selection of complex projections

ayende committed:
Merge branch 'RavenDB-8212' of github.com:ayende/ravendb into v4.0

ayende committed:
RavenDB-8212
- Making sure that we process the order by / where after processing the aliases in the select
- Fixing an issue with bad error if we can't sum a string / null

ayende committed:
RavenDB-8212 - fixing index out of range error

ayende committed:
RavenDB-8212 - Making it easier to tell if we are loading an array or an object with load

aviv86 committed:
RavenDB-7463
- in relevant endpoints, if db is missing add Database-Missing value to the response headers
- RequestExecutor.ExecuteAsync() : throw DatabaseDoesNotExistsException instead of AllTopologyNodesDownException when database is missing

ayende committed:
RavenDB-8212
- Fixing compilation of the project
- Removed a lot of the transformers tests if they didn't add something of value beyond testing transformers functionality (converted some tests / added issues to ensure we retain the capabilities)

grisha-kotler committed:
RavenDB-8230 - skip transformers import when using Smuggler
fix error messages

aviv86 committed:
RavenDB-7463 : fix failing test

ayende committed:
RavenDB-8212 - Fixing queries on the count of a json object

ayende committed:
RavenDB-8212 - Fixing test

ayende committed:
Merge branch 'v4.0' of https://github.com/ravendb/ravendb into v4.0

arekpalinski commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -117,33 +118,92 @@ private void Build(BlittableJsonReaderObject parameters)
+
+        private static void ThrowUseOfReserveFunctionBodyMethodName()
+        {
+            throw new InvalidQueryException("When using select function body, the '__selectOutput' function is reserved");
```
This overload of `InvalidQueryException` does not include the query text and parameters. Its existence might be confusing however we need it to properly deserialize the exception on the client side. 

We have `InvalidQueryException(string message, string queryText, BlittableJsonReaderObject parameters)` to throw on the server.

ayende replied:
Can we mark it as private?

ayende replied:
Fixed

arekpalinski replied:
Yes, actually. See PR: https://github.com/ayende/ravendb/pull/4664

arekpalinski commented on src/Sparrow/StringSegment.cs:
```diff
@@ -193,6 +193,14 @@ public bool StartsWith(string prefix)
 
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool EndssWith(string postfix)
```
Typo: EndssWith -> EndsWith

arekpalinski commented on test/SlowTests/Issues/RavenDB_3647.cs:
```diff
@@ -1,104 +0,0 @@
-        }
-
-        [Fact]
-        public void CanLockIndexes()
```
I don't think this test should be removed - it operates on indexes, not transformers

arekpalinski replied:
Restore candidate

ayende replied:
Done

arekpalinski commented on test/SlowTests/Issues/RavenDB_3647.cs:
```diff
@@ -1,104 +0,0 @@
-        }
-
-        [Fact]
-        public void CanLockIndexes()
```
Also shouldn't we rewrite the tests to use the new approach?

ayende committed:
RavenDB-8212 - Fixing extension setter delegate causing deserialization errors

ayende committed:
RavenDB-8212 - PR Review fixes, method name typo, use proper exception ctor for more detailed reporting.

ayende committed:
RavenDB-8212 - Ignore transformers in smuggler import

arekpalinski committed:
RavendB-8212 Allow to create an instance of exception using private ctor. Prevent from using invalid ctor of InvalidQueryException

ayende committed:
RavenDB-8212 - Fixing test

ayende committed:
RavenDB-8212 - Fixing setting on null id to projections

ayende committed:
RavenDB-8212 - Fixing tests

ayende committed:
Merge branch 'RavenDB-8212' of github.com:ayende/ravendb into v4.0

# Conflicts:
#	test/SlowTests/Bugs/LiveProjections/LiveProjectionOnProducts.cs

ayende committed:
Merge branch 'v4.0' of github.com:ayende/ravendb into v4.0

ayende committed:
RavenDB-8212 - Will not override id from metadata if id is already specified

ayende committed:
RavenDB-8212 - Fixing race condition in test

arekpalinski commented on test/SlowTests/Tests/ResultsTransformer/NullDynamicValuesInTransformerOperation.cs:
```diff
@@ -1,60 +0,0 @@
-
-namespace SlowTests.Tests.ResultsTransformer
-{
-    public class NullDynamicValuesInTransformerOperation : RavenTestBase
```
Restore candidate:  here we are testing   + - / % operators of `DynamicNullObject` and check results return `null` then

ayende replied:
Added

arekpalinski commented on test/SlowTests/Tests/ResultsTransformer/NestingTransformers.cs:
```diff
@@ -1,162 +0,0 @@
-
-namespace SlowTests.Tests.ResultsTransformer
-{
-    public class NestingTransformers : RavenTestBase
```
Restore candidate: one function can be called from another

ayende replied:
No, we already have a way to do function calls and they can recurse if needed.

arekpalinski replied:
OK. We just need a test which verifies that

ayende replied:
http://issues.hibernatingrhinos.com/issue/RavenDB-8201

arekpalinski commented on test/SlowTests/SlowTests/MailingList/Jalchr.cs:
```diff
@@ -1,212 +0,0 @@
-            using (var store = GetDocumentStore())
-            {
-                store.Conventions.CustomizeJsonSerializer = serializer =>
-                               serializer.ReferenceLoopHandling = ReferenceLoopHandling.Ignore;
```
Restore candidate? due to `serializer.ReferenceLoopHandling = ReferenceLoopHandling.Ignore;`

ayende replied:
Now that json.NET is not part of our code, we don't actually need it.

arekpalinski commented on test/SlowTests/MailingList/TransformWithLoad.cs:
```diff
@@ -1,234 +0,0 @@
-                                                   ContactId = c.Id,
-                                                   ContactName = c.Name,
-                                                   ContactDetails = LoadDocument<Detail>(c.DetailIds),
-                                                   MetaVal = MetadataFor(c).Value<string>("Val")
```
Restore candidate: create a test to make sure function can access metadata from a doc

ayende replied:
http://issues.hibernatingrhinos.com/issue/RavenDB-8201

ayende replied:
Actually, `ResultTransformerMetaData` already does that :-)

arekpalinski commented on test/SlowTests/MailingList/TransformWithConversionListener.cs:
```diff
@@ -1,117 +0,0 @@
-        }*/
-
-        [Fact(Skip = "RavenDB-6264")]
-        public void SyncApi()
```
Restore candidate: we got rid of listeners in favor of events. Should we have events before/after conversion as it was used here?

ayende replied:
I don't think so, they can override the convention event and use that instead

arekpalinski replied:
Make sense

arekpalinski commented on test/SlowTests/MailingList/TransformResultsWithProjection.cs:
```diff
@@ -1,173 +0,0 @@
-    public class TransformResultsWithProjection : RavenTestBase
-    {
-        [Fact(Skip = "Missing feature: Spatial")]
-        public void CanGetProjectedResultsWhenUsingTransformWithInMemory()
```
Restore candidate: spatial + transformation function combination in the query

ayende replied:
http://issues.hibernatingrhinos.com/issue/RavenDB-8254

ayende replied:
But I'm not expecting any issues there.

arekpalinski replied:
Yeah, just in case

arekpalinski commented on test/SlowTests/MailingList/Tobias.cs:
```diff
@@ -1,186 +0,0 @@
-{
-    public class Tobias : RavenTestBase
-    {
-        [Fact(Skip = "Missing feature: Collation")]
```
Restore candidate: Missing feature: Collation - not sure what needs to be done regarding this

ayende replied:
This is basically creating a ton of analyzers for no real benefit. Users can just define a custom analyzer on their own.

arekpalinski replied:
Actually this test was passing. There was no need to skip it.

ayende replied:
Restored

arekpalinski commented on test/SlowTests/MailingList/TestIndexAndTransforSubcollections.cs:
```diff
@@ -1,96 +0,0 @@
-                {
-                    var result = session.Query<IndexSubCollectionResult, IndexSubCollection>()
-                        .Customize(customization => customization.WaitForNonStaleResultsAsOfNow()
-                            .SetAllowMultipleIndexEntriesForSameDocumentToResultTransformer(true))
```
Restore candidate: in a modified version, due to `SetAllowMultipleIndexEntriesForSameDocumentToResultTransformer` usage? I'm not sure if that's still relevant

ayende replied:
No, I killed that feature entirely. A function can only return an object, never an array.
This make it simpler for us, and users can send an object with array and operate on that. 
Also makes paging much easier.

arekpalinski commented on test/SlowTests/MailingList/SpatialQueryWithTransformTests.cs:
```diff
@@ -1,195 +0,0 @@
-    public class SpatialQueryWithTransformTests : RavenTestBase
-    {
-        [Fact(Skip = "Missing feature: Spatial")]
-        public void CanQuery()
```
Restore candidate: same as previously, due to spatial usage here

ayende replied:
http://issues.hibernatingrhinos.com/issue/RavenDB-8254

arekpalinski commented on test/SlowTests/MailingList/ResultTransformerMetaData .cs:
```diff
@@ -1,149 +0,0 @@
-        /// So im not sure if the @last-modified meta-field will ever be filled in this case
-        /// </summary>
-        [Fact(Skip = "RavenDB-6264")]
-        public void CreateDataAndQuery()
```
Restore candidate: due to the listener usage - let's rewrite it to use events

ayende replied:
Added

arekpalinski commented on test/SlowTests/MailingList/ReduceAndTransform2.cs:
```diff
@@ -1,144 +0,0 @@
-                    // With transform (will give 1 to 4 results, depending on the weather?)
-                    var results2 = session.Query<PersonnelAll.Mapping, PersonnelAll>()
-                                        .TransformWith<PersonnelTransformer, PersonnelTransformer.Result>()
-                                        .ToArray();
```
Restore candidate: no idea how it could return wrong results due to transformer usage, however we'll be better off having test checking multi map reduce index + query defined transformation

ayende replied:
Added

ayende replied:
Added

arekpalinski commented on test/SlowTests/MailingList/RavenArrayQueryInputTests.cs:
```diff
@@ -1,59 +0,0 @@
-            {
-                TransformResults = documents => from d in documents
-                                                let index = Array.IndexOf(Parameter("array").Value<string>().Split(',', StringSplitOptions.None), d.Id)
-                                                orderby index
```
How to achieve that in RQL? I mean `orderby` previously handled by the transformer using a paramater

ayende replied:
You don't. We won't allow to send more results from the function.
You can send an object, what you do with the object is completely up to you

arekpalinski commented on test/SlowTests/MailingList/DictionaryOfEnum.cs:
```diff
@@ -1,78 +0,0 @@
-
-namespace SlowTests.MailingList
-{
-    public class DictionaryOfEnum : RavenTestBase
```
Restore candidate: make sure we properly deal with dicts of enums

ayende replied:
Added

arekpalinski commented on test/SlowTests/MailingList/CanTransformWithDynamicFields3.cs:
```diff
@@ -1,223 +0,0 @@
-
-namespace SlowTests.MailingList
-{
-    public class CanTransformWithDynamicFields : RavenTestBase
```
Restore candidate: I think this one is worth to be restored since it deals with dynamic fields

ayende replied:
Yes, this actually exposed an issue: RavenDB-8257

arekpalinski commented on test/SlowTests/Issues/RavenDb_2239.cs:
```diff
@@ -1,236 +0,0 @@
-
-namespace SlowTests.Issues
-{
-    public class RavenDb_2239 : RavenTestBase
```
Restore candidate: this test caused request timeout issue originally, also FullLogTransformerDelay test contains an assertion for the steaming duration with transformation applied

ayende replied:
Added

ayende committed:
- Avoid using the same server instance for failover
- Better error on failure to send requests

arekpalinski commented on test/SlowTests/Issues/RavenDB_2793.cs:
```diff
@@ -1,76 +0,0 @@
-                TransformResults = people => from person in people
-                                             select new
-                                             {
-                                                 FirstName = person.FirstName + Parameter("value").Value<int>()
```
Restore candidate: make sure we properly handle numeric parameters in transformation function

ayende replied:
Not needed, we are already sending them via the exact same mechanism as we send query params, so that is covered

arekpalinski commented on test/SlowTests/Issues/RavenDB_2495.cs:
```diff
@@ -1,24 +0,0 @@
-        {
-            using (var x = new MoreLikeThisTests())
-            {
-                x.IncludesShouldWorkWithMoreLikeThis();
```
Restore candidate: no idea why it was done like here, but we need to add [Fact] to `IncludesShouldWorkWithMoreLikeThis` , `TransformersShouldWorkWithMoreLikeThis1` and `TransformersShouldWorkWithMoreLikeThis2` in `MoreLikeThisTests.cs`

arekpalinski replied:
Okay. It was moved to Slow tests while `MoreLikeThisTests.cs` is in FastTests. Needs to be restored.

ayende replied:
http://issues.hibernatingrhinos.com/issue/RavenDB-8229

arekpalinski commented on test/SlowTests/Issues/RavenDB_2496.cs:
```diff
@@ -1,33 +0,0 @@
-
-namespace SlowTests.Issues
-{
-    public class RavenDB_2496 : RavenTestBase
```
Restore candidate: needs to be restored

arekpalinski commented on test/SlowTests/Issues/RavenDB_2435.cs:
```diff
@@ -1,121 +0,0 @@
-                TransformResults = results => from result in results
-                                              select new
-                                              {
-                                                  Name = result.Name.StripHtml()
```
Do we want to offer built-in `stripHtml`?

ayende replied:
I don't think so. They can write a regex there.

arekpalinski commented on test/SlowTests/Issues/RavenDB_2435.cs:
```diff
@@ -1,121 +0,0 @@
-                Map = people => from person in people
-                                select new
-                                {
-                                    Name = person.Name.StripHtml()
```
Restore candidate: it verifies `DynamicExtensionMethods.StripHtml` is working during _indexing_

ayende replied:
I removed this entirely. It should be done by an analyzer, we shouldn't provide it. Note that users can still add via the code addition feature, no need for us to handle this.

arekpalinski commented on test/SlowTests/Issues/RavenDB-5241.cs:
```diff
@@ -1,108 +0,0 @@
-    public class RavenDB_5241 : RavenTestBase
-    {
-        [Fact]
-        public void loading_documents_with_transformer_duplicate_ids()
```
Restore candidate: duplicated ids send to the server, results don't contain duplicates

ayende replied:
added

arekpalinski commented on test/SlowTests/Core/Indexing/ReferencedDocuments.cs:
```diff
@@ -27,65 +26,6 @@ namespace SlowTests.Core.Indexing
     public class ReferencedDocuments : RavenTestBase
     {
         [Fact]
-        public void CanUseLoadDocumentToIndexReferencedDocs()
```
Restore candidate: here we test more indexing of referenced docs that transformers.

arekpalinski commented on test/SlowTests/Core/Indexing/ReferencedDocuments.cs:
```diff
@@ -27,65 +26,6 @@ namespace SlowTests.Core.Indexing
-
-
-                    var companies = session.Advanced.DocumentQuery<Companies_WithReferencedEmployees.CompanyEmployees>(companiesWithEmployees.IndexName)
-                        .SetTransformer(companiesWithEmployeesTransformer.TransformerName)
```
We can just remove this lines and three `Assert.Equal("Last Name x", companies[0].Employees[x]); ` assersions below

ayende replied:
added

arekpalinski commented:
Review of tests which I think should be restored is done.

ayende committed:
RavenDB-8212 - Restored removed test and ensure that null coalescing actually works

ayende committed:
RavenDB-8212 - Restoring test

ayende committed:
RavenDB-8212 - Restoring test and making sure that we can call OnBeforeStore to modify the metadata of the document.

ayende committed:
RavenDB-8212 - Restoring tests, making sure will get good errors if streaming indexes that aren't there.

ayende committed:
RavenDB-8212 - Removing StripHtml. Users that want it can add it back using AdditionalSources in indexing

ayende committed:
RavenDB-8212 restoring tests

ayende committed:
RavenDB-8212 - Fixing modification changes when a listener modified the metadata

ayende committed:
RavenDB-8212 - Fixing tests and removing HtmlAgilityPack

maximburyak committed:
RavenDB-7534: WIP

ayende committed:
Merge branch 'v4.0' of github.com:ayende/ravendb into v4.0

ayende committed:
RavenDB-8212
- Starting to move to Jurrasic and major refactoring
- Dropped custom functions feature

ayende committed:
RavenDB-8212
- More refactoring of how we are executing javascript
- Using a dedicated key to avoid additional allocations if we need to edit the script
- Basic patches work now

ayende committed:
RavenDB-8212 - Allow to pass arguments to patch scripts

ayende committed:
RavenDB-8212 - Subscription scripts now work, better cache key

ayende committed:
RavenDB-8212 - Fixing SQL Etl scripts. Making sure that we have null propagation working properly

ayende committed:
RavenDB-8212 - Fixing Raven ETL scripts, replacing __document_id with id(this)

ayende committed:
RavenDB-8212 - Fixing conflict resolution scripts, we now use `return resolveToTombstone` to indicate a tombstone resolution

ayende committed:
RavenDB-8212 - Moving AdminConsole to Jurrasic as well, we'll not be auto evaluating the script but require explicit action on the cmd line console

ayende committed:
RavenDB-8265 - Ignore Admin Console tests for now, until we figure out how to call them properly.

ayende committed:
RavenDB-8212 - Fixing patch scripts in smuggler

ayende committed:
RavenDB-8212 - Changing the format of smuggler scripts to match the rest of our scripts.

ayende committed:
RavenDB-8212 - Moved queries to use Jurrasic as well

ayende committed:
RavenDB-8266
- Protect against recursion when translating to / from Jurrasic
- Track nested values as well so we can modify them on patch scripts
- Fixing patch by index
- Updating tests for new syntax

ayende committed:
RavenDB-8212 - Fixing tests for patching

ayende committed:
RavenDB-8262 Restore patch output() function and how it is sent to the user

ayende committed:
RavenDB-8212 - More fixes for the patch tests

ayende committed:
RavenDB-8212
- Restoring embedding of document id in nested objects
- Fixing NREs
- Workaround to Jurrasic long issues

ayende committed:
RavenDb-8212 - Don't always embed the @id in, only when we are going to send it out

ayende committed:
RavenDB-8212 - Fixing tests

ayende committed:
RavenDB-8212 - Actually, there is a slow test using this, so we have to keep it this way

ayende committed:
RavenDB-8212 - Enabling collection patching again

ayende committed:
RavenDB-8212 - Fixing tests, avoiding trying to serialize a function instance

ayende committed:
RavenDb-8212 - More tests work

ayende committed:
RavenDB-8268 select Address should not replace full result set
Removing tests for custom functions that we removed
Fixed more tests

ayende committed:
RavenDB-8212 - When putting a document, mark it as written to disk

ayende committed:
RaveDB-8212 - More test fixes

ayende committed:
RavenDB-8212 - More test fixes

ayende committed:
RavenDB-8212 - Guess based on the previous type of the property what type it is (long / double)

ayende committed:
RavenDB-8212 - allow to get lastModified(doc) from scripts

ayende committed:
Avoid crash if timer is running and being disposed at the same time

ayende committed:
- Fixing crash on timer Dispose
- Fixing NRE when trying to match on an index that was just deleted
- More test fixes

ayende committed:
RavenDB-8212 - Fixing output of patch conflict documents

ml054 committed:
RavenDB-7050 remove Custom functions

# Conflicts:
#	test/SlowTests/Issues/RavenDB_3197.cs

ayende committed:
RavenDB-8212 - minor refactoring, removing test code

ayende committed:
RavenDB-8212
- Can't handle reference equality to null propogated value with Jurrasic
- Fixed how we are handling deletes of properties in BlittableObjectInstance
- Fixing tests

ayende committed:
RavenDB-8212 - Fixing passing of null id to a put() method in scripts
